File: server/modules/settings/store.ts
import { eq, like } from "drizzle-orm";
import type { AppDatabase } from "../../db/index.js";
import { settings } from "../../db/schema.js";

export interface SettingsStore {
  get<T>(key: string, fallback: T): Promise<T>;
  set<T>(key: string, value: T): Promise<void>;
  getAll(prefix?: string): Promise<Record<string, unknown>>;
}

export function createSettingsStore(db: AppDatabase["db"]): SettingsStore {
  const cache = new Map<string, unknown>();

  return {
    async get<T>(key: string, fallback: T): Promise<T> {
      if (cache.has(key)) {
        return cache.get(key) as T;
      }
      const rows = await db.select().from(settings).where(eq(settings.key, key)).limit(1);
      const row = rows[0];
      if (!row) return fallback;
      try {
        const parsed = JSON.parse(row.value) as T;
        cache.set(key, parsed);
        return parsed;
      } catch {
        return fallback;
      }
    },

    async set<T>(key: string, value: T): Promise<void> {
      const encoded = JSON.stringify(value);
      const updatedAt = new Date();
      await db
        .insert(settings)
        .values({ key, value: encoded, updatedAt })
        .onConflictDoUpdate({
          target: settings.key,
          set: { value: encoded, updatedAt },
        });
      cache.set(key, value);
    },

    async getAll(prefix?: string): Promise<Record<string, unknown>> {
      const rows = prefix
        ? await db.select().from(settings).where(like(settings.key, `${prefix}%`))
        : await db.select().from(settings);
      const out: Record<string, unknown> = {};
      for (const row of rows) {
        try {
          out[row.key] = JSON.parse(row.value);
        } catch {
          out[row.key] = row.value;
        }
      }
      return out;
    },
  };
}

File: server-rs/src/settings/store.rs
use crate::db::Db;
use crate::error::{AppError, AppResult};
use parking_lot::Mutex;
use rusqlite::params;
use serde_json::Value;
use std::collections::HashMap;
use std::sync::Arc;

pub struct SettingsStore {
    db: Arc<Db>,
    cache: Mutex<HashMap<String, Value>>,
}

impl SettingsStore {
    pub fn new(db: Arc<Db>) -> Self {
        Self {
            db,
            cache: Mutex::new(HashMap::new()),
        }
    }

    pub fn get<T: serde::de::DeserializeOwned + serde::Serialize>(&self, key: &str, fallback: T) -> AppResult<T> {
        if let Some(cached) = self.cache.lock().get(key) {
            if let Ok(value) = serde_json::from_value(cached.clone()) {
                return Ok(value);
            }
        }

        let conn = self.db.conn();
        let mut stmt = conn
            .prepare("SELECT value FROM settings WHERE key = ?1 LIMIT 1")
            .map_err(|e| AppError::Internal(e.to_string()))?;

        let result = stmt.query_row(params![key], |row| row.get::<_, String>(0));

        match result {
            Ok(raw) => match serde_json::from_str::<T>(&raw) {
                Ok(parsed) => {
                    self.cache
                        .lock()
                        .insert(key.to_string(), serde_json::to_value(&parsed).unwrap_or(Value::Null));
                    Ok(parsed)
                }
                Err(_) => Ok(fallback),
            },
            Err(rusqlite::Error::QueryReturnedNoRows) => Ok(fallback),
            Err(e) => Err(AppError::Internal(e.to_string())),
        }
    }

    pub fn set<T: serde::Serialize>(&self, key: &str, value: &T) -> AppResult<()> {
        let encoded = serde_json::to_string(value)
            .map_err(|e| AppError::Internal(format!("Failed to encode setting: {e}")))?;
        let updated_at = chrono::Utc::now().timestamp_millis();

        self.db
            .conn()
            .execute(
                "INSERT INTO settings (key, value, updated_at) VALUES (?1, ?2, ?3)
                 ON CONFLICT(key) DO UPDATE SET value = excluded.value, updated_at = excluded.updated_at",
                params![key, encoded, updated_at],
            )
            .map_err(|e| AppError::Internal(e.to_string()))?;

        self.cache
            .lock()
            .insert(key.to_string(), serde_json::to_value(value).unwrap_or(Value::Null));
        Ok(())
    }

    pub fn get_all(&self, prefix: Option<&str>) -> AppResult<HashMap<String, Value>> {
        let conn = self.db.conn();
        let mut out = HashMap::new();

        if let Some(prefix) = prefix {
            let mut stmt = conn
                .prepare("SELECT key, value FROM settings WHERE key LIKE ?1")
                .map_err(|e| AppError::Internal(e.to_string()))?;
            let rows = stmt
                .query_map(params![format!("{prefix}%")], |row| {
                    Ok((row.get::<_, String>(0)?, row.get::<_, String>(1)?))
                })
                .map_err(|e| AppError::Internal(e.to_string()))?;

            for row in rows {
                let (key, raw) = row.map_err(|e| AppError::Internal(e.to_string()))?;
                out.insert(
                    key,
                    serde_json::from_str(&raw).unwrap_or(Value::String(raw)),
                );
            }
        } else {
            let mut stmt = conn
                .prepare("SELECT key, value FROM settings")
                .map_err(|e| AppError::Internal(e.to_string()))?;
            let rows = stmt
                .query_map([], |row| {
                    Ok((row.get::<_, String>(0)?, row.get::<_, String>(1)?))
                })
                .map_err(|e| AppError::Internal(e.to_string()))?;

            for row in rows {
                let (key, raw) = row.map_err(|e| AppError::Internal(e.to_string()))?;
                out.insert(
                    key,
                    serde_json::from_str(&raw).unwrap_or(Value::String(raw)),
                );
            }
        }

        Ok(out)
    }
}
